pkg/querier/tenantfederation/merge_queryable.go

```diff
@@ -96,6 +97,9 @@ type mergeQueryable struct {
 // Querier returns a new mergeQuerier, which aggregates results from multiple
 // underlying queriers into a single result.
 func (m *mergeQueryable) Querier(ctx context.Context, mint int64, maxt int64) (storage.Querier, error) {
+	// TODO: it's necessary to think how to override context inside querier
+	//  to mark spans created inside querier as child of a span created inside
+	//  methods of merged querier.
```

Comment:
I think it would make sense to create a span during the creation of the queriers and the finish the span once `Close()` of the resulting mergeQuerier is called.

I have played with that here: 

https://github.com/cortexproject/cortex/commit/6ebd273bf665752794843aa02495b8a6ede0f15f

That's how that looks like:
![scrn-2021-05-17-12-46-02](https://user-images.githubusercontent.com/223048/118483403-df13b280-b70d-11eb-95dd-18cb4899fa99.png)

Reply:
This feels at least like it can group the same tenants nicely together. Obvisouly for a real fix we need context propagation as a part of LabelValues,LabelName, Select.

Reply:
yes, it will help to group spans by tenant_id . thanks, i will implement it

Reply:
however, it might confuse a little bit because all spans `mergeQuerier.NewQuerier` will take almost similar time for each tenant. but inside this span we will see real picture. 
in current implementation it looks less confusing, i think. 
![image](https://user-images.githubusercontent.com/82767850/118489929-0eceb480-b726-11eb-96a1-4ef9397c89f8.png)

what do you think?

Reply:
To be honest I agree partly with your point, but I am not too sure, how to fix this properly. 

I guess part of the problem is that the a query is only finished once `Close()` is called, which happens at the same time for underlying queriers.

I think the most confusing part of not grouping spans by tenant_id is seeing all (and even unrelated) spans together. I think until we implement context passing in those methods, we won't be able to solve that properly.

Reply:
You could instantiate the queriers (call the callback inside LabelValues, etc.) and the right context will be applied.